### PR TITLE
chore(ci): Add reporting of no pod restarts for containers

### PR DESCRIPTION
**Pull request by mtodor**
### Description

This PR addresses several test collection inefficiencies.

1. Currently, we are collecting only pod restarts as failures. We don't have "successful Pod Restarts tests". Because of that, we are not able to calculate the failure ratio for a specific pod.
2. Restarts are collected on the Pod level, but restarts are happening for containers, and in some cases, it is difficult to understand who is responsible for failure in cases where we have multiple containers within a pod.
3. Collected restarts are taking only the first string before `-` to report a failed pod. That makes a collection of `central` and `central-db` failures to be grouped under the same ticket.

This PR addresses all listed problems.

**Notes for a reviewer**
- bash associative arrays are used
- maybe we should consider adding tests while taking logic complexity into consideration. I'm not sure if we have tests for CI-related scripts.

### User-facing documentation

- [x] CHANGELOG is updated **OR** update is not needed
- [x] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

### Testing and quality

- [x] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag
- [x] CI results are inspected

#### Automated testing

- [x] added bats tests

#### How I validated my change

I have used an example file:
```
This restart does not match any ignore patterns
collector-12345-node-inventory-previous.log copied to Artifacts
This restart does not match any ignore patterns
central-566dcd968-9jhqf-central-previous.log copied to Artifacts
This restart does not match any ignore patterns
scanner-v4-c89441231-99991-matcher-previous.log copied to Artifacts
This restart does not match any ignore patterns
unknown-pod-c89441231-99991-container-previous.log copied to Artifacts
This restart does not match any ignore patterns
collector-mxsq5-node-inventory-previous.log copied to Artifacts
```

Created dummy functions:
```
save_junit_failure() {
    echo "Fail: $1 $2"
}

save_junit_success() {
    echo "Success: $*"
}
```

And run for - no fails scenario:
```
$ junit_report_pod_restarts 
Success: Pod Restarts pod: central-db - container: init-db
Success: Pod Restarts pod: collector - container: node-inventory
Success: Pod Restarts pod: collector - container: collector
Success: Pod Restarts pod: collector - container: compliance
Success: Pod Restarts pod: scanner-v4-db - container: init-db
Success: Pod Restarts pod: central - container: central
Success: Pod Restarts pod: scanner-v4 - container: matcher
Success: Pod Restarts pod: scanner-v4-db - container: db
Success: Pod Restarts pod: central-db - container: central-db
Success: Pod Restarts pod: config-controller - container: manager
Success: Pod Restarts pod: admission-control - container: admission-control
Success: Pod Restarts pod: scanner-v4 - container: indexer
Success: Pod Restarts pod: sensor - container: sensor
```

And with a fail scenario:
```
$ junit_report_pod_restarts "$(cat example.log)"
Fail: Pod Restarts unknown
Fail: Pod Restarts pod: collector - container: node-inventory
Fail: Pod Restarts pod: central - container: central
Fail: Pod Restarts pod: scanner-v4 - container: matcher
Success: Pod Restarts pod: central-db - container: init-db
Success: Pod Restarts pod: collector - container: collector
Success: Pod Restarts pod: collector - container: compliance
Success: Pod Restarts pod: scanner-v4-db - container: init-db
Success: Pod Restarts pod: scanner-v4-db - container: db
Success: Pod Restarts pod: central-db - container: central-db
Success: Pod Restarts pod: config-controller - container: manager
Success: Pod Restarts pod: admission-control - container: admission-control
Success: Pod Restarts pod: scanner-v4 - container: indexer
Success: Pod Restarts pod: sensor - container: sensor
```

*EDIT:*

After a successful run. I also checked DB:
```
SELECT * FROM `acs-san-stackroxci.ci_metrics.stackrox_tests` 
WHERE Classname LIKE "%Pod Restarts%" AND BuildTag LIKE "%gda49ffe711"
ORDER BY Timestamp DESC
LIMIT 100
```

We have records with a `passed` state.

Also results are visible in prow UI: https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/stackrox_stackrox/13312/pull-ci-stackrox-stackrox-master-gke-nongroovy-e2e-tests/1856354343708528640

### CODE REVIEW

**Review by porridge — COMMENTED** [review]
Some questions and suggestions inline.

It would be great if there were some unit tests for this code, if only to show what the intended behaviour is in various corner cases 🤔

File: tests/e2e/lib.sh [hunk 1]
```diff
@@ -665,6 +665,83 @@ check_for_stackrox_OOMs() {
...
+    POD_CONTAINERS_MAP["pod: collector - container: collector"]="collector-[A-Za-z0-9]+-collector-previous.log"
+    POD_CONTAINERS_MAP["pod: collector - container: compliance"]="collector-[A-Za-z0-9]+-compliance-previous.log"
+    POD_CONTAINERS_MAP["pod: collector - container: node-inventory"]="collector-[A-Za-z0-9]+-node-inventory-previous.log"
+}
+
+junit_report_pod_restarts() {
+    local check_out="${1:-}"
+
+    local previous_logs=()
+    while IFS='' read -r line; do previous_logs+=("$line"); done < <(echo "${check_out}" | grep "copied to Artifacts" | cut -d" " -f1 | sort -u)
```
Comment (vikin91) [inline tests/e2e/lib.sh:694]: Please add a comment in the place where this log line is emitted saying that it should not be changed as there are scripts that rely on it.

File: tests/e2e/lib.sh [hunk 2]
```diff
@@ -665,6 +665,83 @@ check_for_stackrox_OOMs() {
...
+
+    fill_pod_containers_map
```
[the remark this answers is not in the document]
Comment (mtodor) [reply tests/e2e/lib.sh:696]: Changed to the global map and init on script load.

File: tests/e2e/lib.sh (hunk 1 above)
Comment (mtodor) [reply tests/e2e/lib.sh:694]: Comment is already there: https://github.com/stackrox/stackrox/blob/7e49062da60fbe153d811e42dbcedf8df10bef5a/scripts/ci/logcheck/check-restart-logs.sh#L51

File: tests/e2e/lib.sh [hunk 3]
```diff
@@ -665,6 +665,83 @@ check_for_stackrox_OOMs() {
     fi
 }
 
+declare -A POD_CONTAINERS_MAP
+fill_pod_containers_map() {
+    POD_CONTAINERS_MAP=()
+
+    POD_CONTAINERS_MAP["pod: central - container: central"]="central-[A-Za-z0-9]+-[A-Za-z0-9]+-central-previous.log"
```
[the remark this answers is not in the document]
Comment (mtodor) [reply tests/e2e/lib.sh:672]: I was thinking about this also. Another aspect is that we have different pods for different test runs. (i.e. scanner-v4 runs only on scanner-v4 tests, operator runs only on openshift tests, etc.)

My thought was to **update the list when we face failure**.

Right now, tickets for "known" containers will be:
```
Pod Restarts pod: collector - container: node-inventory
```
Where for "unknown" it will be:
```
Pod Restarts collector
```
We should simply update our list if we see such tickets.

File: tests/e2e/lib.sh [hunk 4]
```diff
@@ -665,6 +684,60 @@ check_for_stackrox_OOMs() {
...
+    # Report pods without restarts (SUCCESSES).
+    local found_map_key=""
+    for map_key in "${!POD_CONTAINERS_MAP[@]}"
+    do
+        found_map_key="${found_pod_container_keys[${map_key}]:-}"
+        if [[ "${found_map_key}" == "found" ]]; then
+            continue
+        fi
+
+        save_junit_success "Pod Restarts" "${map_key}"
```
Comment (vikin91) [inline tests/e2e/lib.sh:737]: This sounds a bit strange 😅 
```
$ junit_report_pod_restarts 
Success: Pod Restarts pod: central-db - container: init-db
Success: Pod Restarts pod: collector - container: node-inventory
Success: Pod Restarts pod: collector - container: collector
(...)
```
it looks like we would want the pods to restart and that operation was successful. 

Maybe we could use something like:
```suggestion
        save_junit_success "Check unexpected pod restarts" "${map_key}"
```
then status success and failure would make sense. WDYT?

**Review by vikin91 — APPROVED** [review]
I gave it another look and it looks good. I left one comment on the wording of the output - maybe we could avoid potential confusions?

File: tests/e2e/lib.sh (hunk 4 above)
Comment (vikin91) [reply tests/e2e/lib.sh]: Thanks for improoving!
